### ISSUE TITLE
use_transactions: false not respected in execution

### ISSUE THREAD

**Issue by rosskevin** [state-machines/state_machines-activerecord#16]
I should note that I am debugging this in a `before_transition`.

I'm not sure what is going on yet, but I can see that the [machine.rb#within_transaction](https://github.com/state-machines/state_machines/blob/master/lib/state_machines/machine.rb#L1862) is not run (which seems to evaluate this flag).   With the active record integration, it runs the transaction every time.  I'm not sure if it is overwriting that check in machine or what is occuring, but I cannot seem to turn off transactions as [this code](https://github.com/state-machines/state_machines-activerecord/blob/master/lib/state_machines/integrations/active_record.rb#L536) is always run:

```ruby
      # Runs a new database transaction, rolling back any changes by raising
      # an ActiveRecord::Rollback exception if the yielded block fails
      # (i.e. returns false).
      def transaction(object)
        result = nil
        object.class.transaction do
          raise ::ActiveRecord::Rollback unless result = yield
        end
        result
      end
```

The state_machine is configured with:

```ruby
    state_machine :state, initial: :uninitialized, use_transactions: false do
```

**Comment by rosskevin** [comment]
[`#around_save`](https://github.com/state-machines/state_machines-activerecord/blob/master/lib/state_machines/integrations/active_record.rb#L509) appears to be the culprit.  No check for the flag, just use the transaction.  It may not be checking for the flag because there could be one or more state_machines in this block:

```ruby
      # Runs state events around the machine's :save action
      def around_save(object)
        transaction(object) do
          object.class.state_machines.transitions(object, action).perform { yield }
        end
      end
```

**Comment by rosskevin** [comment]
The mongoid integration [`#around_save`](https://github.com/state-machines/state_machines-mongoid/blob/master/lib/state_machines/integrations/mongoid.rb#L376) is simpler, wondering if it indicates that the `active_record#around_save` is coded incorrectly?

```ruby
      # Runs state events around the machine's :save action
      def around_save(object)
        object.class.state_machines.transitions(object, action).perform { yield }
      end
```

**Comment by rosskevin** [comment]
I'll try to get back to this in a few hours.  @seuros if you could give me your opinion on what I've found, I'll pick it back up and create a pull request.

**Comment by rosskevin** [comment]
So who is responsible for interpreting the transaction option?  StateMachine or TransitionCollection?  

TransitionCollection appears to get into the interpreting of transactions and even has an [errant hack](https://github.com/state-machines/state_machines/blob/master/lib/state_machines/transition_collection.rb#L181) that picks the first state_machine and uses transactions (problem 1), but only if [**it** is configured with `transaction: true`](https://github.com/state-machines/state_machines/blob/master/lib/state_machines/transition_collection.rb#L36) for which in my execution it is [hardcoded with `transaction: false`](https://github.com/state-machines/state_machines/blob/master/lib/state_machines/transition_collection.rb#L195).  On top of this, the [active_record#around_save](https://github.com/state-machines/state_machines-activerecord/blob/master/lib/state_machines/integrations/active_record.rb#L510) forces transactions with no regard for either the TransitionCollection's `transactions` option or the state_machine's `use_transactions` option.

So here is what I see (and a proposal):
1.  StateMachine is coded properly, and should be the interpreter of `use_transactions`.  If TransitionCollection needs access to the option for interpretation, it should come from (in order of priority) 1. transition's `:transaction` option, or 2. state_machine's `use_transaction`, or 3. default to hardcoded.
2.  TransitionCollection is [improperly making decisions](https://github.com/state-machines/state_machines/blob/master/lib/state_machines/transition_collection.rb#L181) about transactions and it should be removed (delegated to state_machine#within_transaction)
3. AttributeTransitionCollection (subclass) is [hardcoded with `transaction: false`](https://github.com/state-machines/state_machines/blob/master/lib/state_machines/transition_collection.rb#L195) and should be given the same treatment as TransitionCollection specified above.
4. [active_record#around_save](https://github.com/state-machines/state_machines-activerecord/blob/master/lib/state_machines/integrations/active_record.rb#L510) should not interpret transactions, this is the job of `StateMachine#within_transaction`
5. active_record.rb should only need to implement `#transaction` for the underlying ORM, which [it already does](https://github.com/state-machines/state_machines-activerecord/blob/master/lib/state_machines/integrations/active_record.rb#L536)

This would be 2 pull requests because it will cross-project boundaries, but I think it's the correct/intended set of responsibilities.

@seuros - your thoughts?

**Comment by rosskevin** [comment]
I'm working on the pull's now, it was a bit simpler than I thought, but should cure the inconsistencies mentioned.

**Comment by rosskevin** [comment]
This is solved with the two pull requests, see #17 for details.

@seuros if there is anything you see wrong, please let me know.
